controller/src/main/java/io/pravega/controller/store/stream/records/TruncationRecord.java

```diff
@@ -0,0 +1,177 @@
+import io.pravega.common.io.serialization.VersionedSerializer;
+import lombok.Builder;
+import lombok.Data;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.DataInput;
+import java.io.DataOutput;
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Data class for storing information about stream's truncation point.
+ */
+@Data
+@Slf4j
+public class TruncationRecord {
+    public static final TruncationRecordSerializer SERIALIZER = new TruncationRecordSerializer();
+
+    public static final TruncationRecord EMPTY = new TruncationRecord(ImmutableMap.of(),
+            ImmutableMap.of(), ImmutableSet.of(), ImmutableSet.of(), false);
+
+    /**
+     * Stream cut that is applied as part of this truncation.
+     */
+    private final ImmutableMap<Long, Long> streamCut;
+
+    /**
+     * If a stream cut spans across multiple epochs then this map captures mapping of segments from the stream cut to
+     * epochs they were found in closest to truncation point.
+     * This data structure is used to find active segments wrt a stream cut.
+     * So for example:
+     * epoch 0: 0, 1
+     * epoch 1: 0, 2, 3
+     * epoch 2: 0, 2, 4, 5
+     * epoch 3: 0, 4, 5, 6, 7
+     *
+     * Following is a valid stream cut {0/offset, 3/offset, 6/offset, 7/offset}
+     * This spans from epoch 1 till epoch 3. Any request for segments at epoch 1 or 2 or 3 will need to have this stream cut
+     * applied on it to find segments that are available for consumption.
+     * Refer to TableHelper.getActiveSegmentsAt
+     */
+    private final ImmutableMap<StreamSegmentRecord, Integer> span;
+
+    /**
+     * All segments that have been deleted for this stream so far.
+     */
+    private final ImmutableSet<Long> deletedSegments;
+    /**
+     * Segments to delete as part of this truncation.
+     * This is non empty while truncation is ongoing.
+     * This is reset to empty once truncation completes by calling mergeDeleted method.
+     */
+    private final ImmutableSet<Long> toDelete;
+
+    private final boolean updating;
+
+    @Builder
+    public TruncationRecord(Map<Long, Long> streamCut, Map<StreamSegmentRecord, Integer> span,
+                            Set<Long> deletedSegments, Set<Long> toDelete, boolean updating) {
+        this.streamCut = ImmutableMap.copyOf(streamCut);
+        this.span = ImmutableMap.copyOf(span);
+        this.deletedSegments = ImmutableSet.copyOf(deletedSegments);
+        this.toDelete = ImmutableSet.copyOf(toDelete);
+        this.updating = updating;
+    }
+
+    int getTruncationEpochLow() {
+        return span.values().stream().min(Comparator.naturalOrder()).orElse(Integer.MIN_VALUE);
+    }
+
+    int getTruncationEpochHigh() {
+        return span.values().stream().max(Comparator.naturalOrder()).orElse(Integer.MIN_VALUE);
+    }
+
+    public ImmutableMap<Long, Long> getStreamCut() {
```

Comment:
`ImmutableMap` is a class. I really prefer exposing this as `Map` since it reduces coupling in the callers.